clc/modules/object-storage/src/main/java/com/eucalyptus/objectstorage/ObjectStorageGateway.java

```diff
@@ -2013,113 +2055,317 @@ public DeleteBucketTaggingResponseType deleteBucketTagging(DeleteBucketTaggingTy
+        MalformedXMLException ex = new MalformedXMLException(bucketName);
+        ex.setMessage("More than " + MAX_CORS_RULES + " rules in CORS configuration not allowed");
+        throw ex;
+      }
+
+      // Make sure names are unique and <= 255 chars.
+      HashSet<String> uniqueRuleIds = new HashSet<String>();
+      String ruleId = null;
+      InvalidArgumentException invArgEx = null;
+      
+      for (CorsRule rule : corsConfig.getRules()) {
+        if (rule == null) {
+          break;
+        }
+        ruleId = rule.getId();
+        if (ruleId != null && ruleId.length() > 0) {
+          if (ruleId.length() > 255) {
             invArgEx = new InvalidArgumentException(ruleId);
-            invArgEx.setMessage("RuleId must be unique. Found same ID for more than one rule");
+            invArgEx.setMessage("CORS RuleId > 255 characters");
             throw invArgEx;
           }
-          if (ruleId.length() > 255) {
+          boolean unique = uniqueRuleIds.add(ruleId);
+          if (!unique) {
             invArgEx = new InvalidArgumentException(ruleId);
-            invArgEx.setMessage("RuleId > 255 characters");
+            invArgEx.setMessage("RuleId must be unique. Saw this rule more than once: " + ruleId);
             throw invArgEx;
+            
           }
         }
-        savedRuleIds.add(ruleId);
       }
-      goodRules.add(rule);
-    }
 
-    try {
-      BucketCorsManagers.getInstance().addCorsRules(goodRules, bucket.getBucketUuid());
+      BucketCorsManagers.getInstance().addCorsRules(corsConfig.getRules(), bucket.getBucketUuid());
+
+    } catch (S3Exception s3e) {
+      LOG.warn("Caught S3Exception while setting the CORS configuration for bucket <" + 
+          bucketName + ">, CorrelationId: " + Contexts.lookup().getCorrelationId() + 
+          ", responding to client with: ", s3e);
+      throw s3e;
     } catch (Exception ex) {
-      LOG.error("Caught exception while setting the CORS configuration for bucket <" + bucketName + ">, error: " + ex.getMessage());
-      throw new InternalErrorException(bucketName);
+      LOG.warn("Caught general exception while setting the CORS configuration for bucket <" + 
+          bucketName + ">, CorrelationId: " + Contexts.lookup().getCorrelationId() + 
+          ", responding to client with 500 InternalError because of: ", ex);
+      throw new InternalErrorException(bucketName, ex);
     }
 
     return response;
-
   }
 
   @Override
   public DeleteBucketCorsResponseType deleteBucketCors(DeleteBucketCorsType request) throws S3Exception {
-    Bucket bucket = getBucketAndCheckAuthorization(request);
-    DeleteBucketCorsResponseType response = request.getReply();
+    String bucketName = "nullBucket";
+    DeleteBucketCorsResponseType response = null;
+
+    // Let most NullPointerExceptions be handled by the catch block.
     try {
+      Bucket bucket = getBucketAndCheckAuthorization(request);
+      bucketName = request.getBucket();
```

Comment:
Because the bucket name is used inside the catch block, and I wanted to have a real string there in case it doesn't get initialized before entering the catch block.